src/main/java/org/cyclonedx/gradle/utils/EnvironmentUtils.java

```diff
@@ -0,0 +1,129 @@
+        }
+
+        // GitLab CI
+        String ciProjectUrl = System.getenv("CI_PROJECT_URL");
+        String ciJobId = System.getenv("CI_JOB_ID");
+        if (!isNullOrBlank(ciProjectUrl) && !isNullOrBlank(ciJobId)) {
+            return String.format("%s/-/jobs/%s", ciProjectUrl, ciJobId);
+        }
+
+        // Jenkins
+        String buildUrl = System.getenv("BUILD_URL");
+        if (!isNullOrBlank(buildUrl)) {
+            return buildUrl;
+        }
+        //        String jobUrl = System.getenv("JOB_URL");
+        //        if (!isNullOrBlank(jobUrl)) {
+        //            return jobUrl;
+        //        }
+
+        // Travis CI
+        String travisBuildUrl = System.getenv("TRAVIS_BUILD_WEB_URL");
+        if (!isNullOrBlank(travisBuildUrl)) {
+            return travisBuildUrl;
+        }
+        //        String travisJobUrl = System.getenv("TRAVIS_JOB_WEB_URL");
+        //        if (!isNullOrBlank(travisJobUrl)) {
+        //            return travisJobUrl;
+        //        }
+
+        // CircleCI
+        String circleBuildUrl = System.getenv("CIRCLE_BUILD_URL");
+        if (!isNullOrBlank(circleBuildUrl)) {
+            return circleBuildUrl;
+        }
+
+        // Drone by Harness
+        String droneBuildUrl = System.getenv("DRONE_BUILD_LINK");
+        if (!isNullOrBlank(droneBuildUrl)) {
+            return droneBuildUrl;
+        }
+        return null;
+    }
+
+    /**
+     * Get the URI of the current build from the specified pattern and environment variables and pattern. An example
+     * pattern could be "${SERVER}/jobs/${JOB_ID}". The environment variables are specified as a list of names.
+     * @param pattern the pattern to use to build the URI
+     * @return the URI of the current build or null if it cannot be determined
+     */
+    private static String getBuildUriFromPattern(String pattern) {
+        String result = pattern;
+        Pattern regex = Pattern.compile("\\$\\{([^}]+)\\}");
+        Matcher matcher = regex.matcher(pattern);
+
+        while (matcher.find()) {
+            String varName = matcher.group(1);
+            if (isNullOrBlank(varName)) {
+                return null;
+            }
+            String value = System.getenv(varName);
+            if (isNullOrBlank(value)) {
+                return null;
+            }
+            result = result.replace("${" + varName + "}", value);
+        }
+        return result;
+    }
+
+    /**
+     * Get the URI of the current build from the specified environment variable.
+     * Additionally, a pattern can be provided to extract the URI from the environment variable.
+     * An example pattern could be "${SERVER}/jobs/${JOB_ID}".
+     *
+     * @param str the name of the environment variable or pattern to use
+     * @return the URI of the current build or null if it cannot be determined
+     */
+    public static String getBuildURI(String str) {
```

Comment:
This method is called before `getBuildUriFromPattern` and is `public`, I think you should define this method before `private static String getBuildUriFromPattern(String pattern) {`.

Also `isNullOrBlank` does not contribute to the public interface and is only a helper method, it could be listed last in the file

Reply:
Thanks, I'll remove the `IsNullOrBlank()` - I missed that commons lang3 is on the classpath.